### PR TITLE
ENH: Add an axis parameter to DataFrame.diff

### PR DESCRIPTION
**Pull request by evanpw**

### CODE REVIEW

**Comment by TomAugspurger** [comment]
+1 on this. I think the axis should be flipped though, right? `axis=0` is a difference along the index, i.e. `df.iloc[1] - df.iloc[0]`. Similar to how `cumsum` operates along the index.

See also `np.diff`.

File: doc/source/whatsnew/v0.16.1.txt [hunk 1]
```diff
@@ -16,7 +16,7 @@ We recommend that all users upgrade to this version.
 
 Enhancements
 ~~~~~~~~~~~~
-
+- ``DataFrame.diff`` now takes an ``axis`` parameter
```
Comment (jreback) [inline doc/source/whatsnew/v0.16.1.txt]: pls add the issue number (which is the PR number, e.g. 9727)

File: pandas/tests/test_frame.py [hunk 2]
```diff
@@ -9955,6 +9955,11 @@ def test_diff_float_n(self):
         xp = self.tsframe.diff(1)
         assert_frame_equal(rs, xp)
 
+    def test_diff_axis(self):
+        df = DataFrame([[1., 2.], [3., 4.]])
```
Comment (jreback) [inline pandas/tests/test_frame.py]: add the issue number here

**Comment by evanpw** [comment]
@TomAugspurger: You're right about ``np.diff`` having a different convention. All of the internal diff machinery (in core.internals.Block, core.common, and C implementations in algos) has this opposite convention as well, but maybe I'll just invert the axis in this public facing ``DataFrame.diff`` method.

**Comment by jreback** [comment]
@evanpw you need to pass the axis as ``self._get_block_manager_axis(axis)`` which handles this. frames (only) are represented with inverted axes.

**Comment by evanpw** [comment]
Ahh, that explains it.

**Comment by jreback** [comment]
@evanpw thanks!

**Pull request merged**